crates/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -105,20 +110,46 @@ impl Context {
         // Otherwise, try inserting the instruction again to apply any optimizations using the newly resolved inputs.
         let new_results = Self::push_instruction(id, instruction.clone(), &old_results, block, dfg);
 
-        // If the instruction is pure then we cache the results so we can reuse them if
-        // the same instruction appears again later in the block.
-        if instruction.is_pure(dfg) {
-            instruction_result_cache.insert(instruction, new_results.clone());
-        }
         Self::replace_result_ids(dfg, &old_results, &new_results);
+
+        Self::cache_instruction(
+            instruction,
+            new_results,
+            dfg,
+            instruction_result_cache,
+            constrained_values,
+        );
     }
 
     /// Fetches an [`Instruction`] by its [`InstructionId`] and fully resolves its inputs.
-    fn resolve_instruction(instruction_id: InstructionId, dfg: &DataFlowGraph) -> Instruction {
+    fn resolve_instruction(
+        instruction_id: InstructionId,
+        dfg: &DataFlowGraph,
+        constrained_values: &mut HashMap<ValueId, ValueId>,
+    ) -> Instruction {
         let instruction = dfg[instruction_id].clone();
 
+        // Alternate between resolving `value_id` in the `dfg` and checking to see if the resolved value
+        // has been constrained to be equal to some simpler value in the current block.
+        //
+        // This allows us to reach a stable final `ValueId` for each instruction input as we add more
+        // constraints to the cache.
+        fn resolve_cache(
+            dfg_resolver: impl Fn(ValueId) -> ValueId,
```

Comment:
Why does this need to be a parameter? It looks like its only value is `|v| dfg.resolve(v)`

Reply:
We need a parameter here as `resolve_cache` can't capture its environment and we can't make it a closure as we need to recurse.

I've switched this over to just pass in `dfg` directly however.